In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score, confusion_matrix

In [2]:
df = sns.load_dataset("titanic")
df = df.dropna(subset=["age", "pclass", "survived"])

print(df["survived"].value_counts())
print(df["survived"].value_counts(normalize=True))

survived
0    424
1    290
Name: count, dtype: int64
survived
0    0.593838
1    0.406162
Name: proportion, dtype: float64


In [3]:
X = df[["age", "pclass"]]
y = df["survived"]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

print(X_train.shape, X_test.shape)
print(y_train.value_counts(normalize=True))
print(y_test.value_counts(normalize=True))

(571, 2) (143, 2)
survived
0    0.593695
1    0.406305
Name: proportion, dtype: float64
survived
0    0.594406
1    0.405594
Name: proportion, dtype: float64


In [4]:
model = LogisticRegression()
model.fit(X_train, y_train)

predictions = model.predict(X_test)

In [5]:
acc = accuracy_score(y_test, predictions)
prec = precision_score(y_test, predictions)
rec = recall_score(y_test, predictions)
f1 = f1_score(y_test, predictions)
probabilities = model.predict_proba(X_test)[:, 1]
roc_auc = roc_auc_score(y_test, probabilities)

print("Accuracy:", acc)
print("Precision:", prec)
print("Recall:", rec)
print("F1:", f1)
print("ROC-AUC:", roc_auc)

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, predictions))

Accuracy: 0.6783216783216783
Precision: 0.6363636363636364
Recall: 0.4827586206896552
F1: 0.5490196078431373
ROC-AUC: 0.7312373225152129

Confusion Matrix:
[[69 16]
 [30 28]]


# Classification Assignment: Predicting Titanic Survival

**Task:** Predict `survived` (0/1) from `age` and `pclass`, using Logistic Regression.

**Class balance:** 59.4% did not survive, 40.6% survived a moderate imbalance, 
much less extreme than the 90/10 split used in the lecture's Carvana example.

**Results (held-out 20%, 143 passengers):**
- Accuracy: 0.678
- Precision: 0.636
- Recall: 0.483
- F1: 0.549
- ROC-AUC: 0.731

**Which metric I'd optimize: Recall**

If this model were used to help prioritize who needs urgent attention in a 
crisis (like a rescue effort), missing an at risk person (a false negative) 
is far more costly than a false alarm. I'd rather overflag some people who 
turn out fine than fail to flag someone who genuinely needed help the same 
logic the lecture used for disease screening.

**Honest limitation:** at 48.3% recall, this model currently misses over half 
of actual survivors, meaning it isn't reliable enough yet for that use case. 
Improving recall even at some cost to precision would be the priority 
before trusting this model in a real prioritization scenario.